lib/gateway/make-resolver.js

```diff
@@ -503,11 +505,46 @@ function makeResolver ({ service, createOperation, transformData, isQuery, isRef
         context
       })
 
-      return transformData(response)
+      const transformed = transformData(response)
+      const transformedTypeName = Array.isArray(transformed) ? transformed[0].__typename : transformed.__typename
+      if (typeToServiceMap) {
+        // If the type is defined in the typeToServiceMap, we need to resolve the type if the type is a reference
+        // and it is fullfilled by another service
+        const targetService = typeToServiceMap[transformedTypeName]
+        // targetService can be null if it is a value type or not defined anywhere
+        if (targetService && targetService !== service.name) {
+          selections = collectServiceTypeFields(fieldNodes[0].selectionSet.selections, serviceMap[targetService], type, schema)
+
+          const toFill = Array.isArray(transformed) ? transformed : [transformed]
+
+          await Promise.all(toFill.map(async transformed => {
+            operation = createEntityReferenceResolverOperation({
+              returnType: transformedTypeName,
+              selections,
+              variableDefinitions: []
+            })
+
+            query = print(operation)
+
+            variables.representations = [{
+              ...removeNonIdProperties(transformed, schema.getType(transformedTypeName)),
```

Comment:
If the `removeNonIdProperties` result is an empty json, could we skip the entity resolver resolution?

For example, does the following query need to be fetched in any case to let the `__resolveReference` does the process?

```gql
meDirect { id }
```

Reply:
Now that I look into it I'm not 100% convinced that removeNonIdProperties is the actual way to implement that behavior. However it would totally be useful and I will open up a follow-up issue to discuss it.